Feature/wasm example

riverKanies opened the pull request:
Here's the quickstart example in WASM.

The code could use some cleaning but I figured I'd go ahead and get some feedback on the direction here.

Seemed like the most important thing was to highlight some of the limitations of using BDK in wasm

UPDATE:
closer to full wallet than quickstart because wanted to include persistence as it is different in wasm

darioAnongba commented:
Thanks so much @riverKanies !

@notmandatory I got the retry mechanism working but might be a bit too obscure as we dive into the `unsafe`: https://github.com/bitcoindevkit/bdk/issues/1683#issuecomment-2489270280

riverKanies commented on docs/cookbook/wasm.md:
````diff
@@ -0,0 +1,77 @@
+--8<-- "examples/wasm/rust/src/lib.rs:new"
+```
+
+Notice we are using an in-memory wallet with `.create_wallet_no_persist()`. If you try to use persistence through file or database you will get an error becuase those features require OS access. Instead we have to create a binding to pass the wallet data to the JavaScript environment where we can handle persistence.
````
great! yes this is one of the main areas I'm wanting to detail more

riverKanies commented:
I believe this is ready for a final review @notmandatory @thunderbiscuit @darioAnongba 

The intention is just to be a minimal example of a custom WASM package for bdk. 
We'll have a separate example for using the bdk-wasm package out of the box at some point (hopefully)

riverKanies commented on examples/wasm/rust/src/lib.rs:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    // --8<-- [start:store]
+    pub fn take_staged(&mut self) -> JsResult<String> {
```
@darioAnongba see use of serde_json here. note I'm converting to json string in this example rn because I'm just saving to localStorage, but you can pas JsValue json and stringify on js side

notmandatory commented:
@riverKanies I tried testing this out today and wasn't able to get it to work. First off it needs some setup instructions in the README since not everyone (including me) has `npm` and `webpack` and `webpack-dev-server` installed. 

But more importantly when I updated to `bdk_wallet` 1.1 and `bdk_esplora` 0.20.1 I couldn't get it to build. I think you need to copy and use the `WebSleeper` that @darioAnongba created for `bdk-wasm`. https://github.com/MetaMask/bdk-wasm/blob/main/src/bitcoin/esplora_client.rs

riverKanies commented:
@notmandatory awsome! thanks so much for taking a look and improving it!

I have rebased and made some minor improvements to readme and index.html
(with regards to webpack, I added a note to `npm install` which should get what you need there)

If you think it's ready to merge I can squash commits

notmandatory commented on docs/cookbook/wasm.md:
```diff
@@ -0,0 +1,77 @@
+Any function that requires system time, such as any sort of timestamp, must access system time through a wasm binding to the JavaScript environment. This means some BDK functions that are commonly used in rust won't work in WASM and instead an alternate rust function that takes a timestamp as an argument must be used (I.E. instead of `.apply_update()` we must use `.apply_update_at()`).
+
+### Network access is limited to http(s)
+This effectively means that the blockchain client must be an Esplora instance. Both RPC and Electrum clients require sockets and will not work for BDK in a WASM environment out of the box.
```
Sorry for the long delay getting back to you on this. In theory yes it is possible to communicate with bitcoind over HTTP but since we're using the `bitcoincore-rpc` crate it only support the RPC over tcp method. If there were some sort of `bitcoincore-http` crate out there or someone builds one then we could support talking to a core node from WASM. Or could just use `minreq` or `reqwest` directly but would be better to have a dedicated crate so each project doesn't need to reinvent the logic.

notmandatory reviewed (approved):
ACK 1955627c56a77e6c47cf42e46cf57d8713ec187b

Looks good, thanks for getting this added!
